bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/Resource.java

```diff
@@ -202,9 +205,19 @@ public State getBrightnessState() {
         if (Objects.nonNull(dimming)) {
             try {
                 // if off the brightness is 0, otherwise it is dimming value
+                // if on and brightness is 0, it is assumed as 100
+                // (because of contradiction, on takes precedence)
                 OnState on = this.on;
-                double brightness = Objects.nonNull(on) && !on.isOn() ? 0f
-                        : Math.max(0f, Math.min(100f, dimming.getBrightness()));
+                double brightness = Math.max(0f, Math.min(100f, dimming.getBrightness()));
+                if (on != null) {
```

Comment:
@jlaur I think I may have misunderstood what you said in the Issue, so let me just check it here..

When the binding starts, it always does an initial download of the full state of all resources. And then the state is updated by delta via individual SSE messages.

Therefore I would assume that even if a light is initially off, the initial full state download would include the last known Dimming state of the lamp. i.e. the full state plus delta should never actually have the case where Dimming == null. => Are you saying that you have actually seen such cases of full plus delta Dimming == null in real life?

Reply:
> Are you saying that you have actually seen such cases of full plus delta Dimming == null in real life?

Actually, there is no change regarding this, and no, I haven't seen that.

But I have seen this in an initial state response just after turning the dimmer on at low dimming by physical button:

`GET https://hue.local/clip/v2/resource/light/00000000-0000-0000-0000-000000000001`

```json
            "on": {
                "on": true
            },
            "dimming": {
                "brightness": 0.0
            },
```

Interesting for this case is that it is shown as 1% brightness in the Hue app. With the fix provided here, we will now interpret as 100% (rather than off/0% as before). This works better for the case I originally detected where the dimmer is turned fully on and the events arrive separately as "on" and "100%", because it will then only cause one state update. But as we discussed in the issue, there is no really good choice anyway, because we simply don't know the actual value.

> .. and what about the case of On/Off only lamps?

They don't have a brightness channel, and the code in unchanged, so it should work fine?

Reply:
These two testcases are failing without the fix, all the others pass in both versions:
- getBrightnessStateWhenOnAndDimming0ReturnBrightness100
- getBrightnessStateWhenOnAndDimmingMinus1ReturnBrightness100

Reply:
>                "brightness": 0.0
> that it is shown as 1% brightness in the Hue app
> With the fix provided here, we will now interpret as 100%
> we simply don't know the actual value

Hmm. I think, therefore, that we should interpret it exactly as the Hue App interprets it. Namely On @ Dimming=0% should be interpreted as On @ Dimming=minimum dimming level.

> it will then only cause one state update

And .. the merging of consecutive SSE events (in this PR) should mean that the first On @ Dimming=minimum dimming level in resource #1 would then be immediately resolved to On @ Dimming=whatever SSE value from resource #2. Or??

Reply:
>> Hmm. I think, therefore, that we should interpret it exactly as the Hue App interprets it. Namely On @ Dimming=0% should be interpreted as On @ Dimming=minimum dimming level.

> There is no minimum dimming level provided by this device (at least which is compatible with Hue), so it's quite a corner case.

>> it will then only cause one state update

> And .. the merging of consecutive SSE events (in this PR) should mean that the first On @ Dimming=minimum dimming level in resource 1 would then be immediately resolved to On @ Dimming=whatever SSE value from resource 2. Or??

I think you are mixing the scenarios? The scenario we are discussing here, is when receiving separate events, i.e. it's not related to merging. So **with** the merge scenario fixed, the scenario would play out like this if we would set 1% (hardcoded, because there is not minimum dimming level):
- Initial state: off, 0%.
- Dimmer is turned on by button directly on device - at 100% brightness.
- We receive on and set brightness to 1%.
- We receive 100% and set brightness to 100%.

So in that case there would be two updates.

The other scenario is:
- Initial state: off, 0%.
- Dimmer is turned on by button directly on device - at low brightness.
- We receive on and set brightness to 1%.

With current code we would set 100% in both scenarios. And in both scenarios the brightness level as unknown, and we can interpret that in any way we see fit.

I don't feel strongly for any of the options, but I tend to slightly prefer 100% because in my most common scenario that would avoid an additional update.

The last scenario with low dimming without any dimming information is probably quite uncommon and device-specific. Did you ever see this?

Reply:
> > I don't care if shown as 1% or 100%, as long as it's on,
> 
> I am Ok with the "as long as it's on" part. However I care that it should be as close to the Hue app as possible. => i.e. minimum dimming level :) .. and not 100% to just suit your scenario, and thus upset mine :(

I didn't know you had a scenario/issue - can you provide some messages documenting this, so we have more information available for making such decision? I'm under the impression that I'm really solving a corner case scenario for a device not behaving well with Hue/Zigbee standard (the hardware is Tuya-based). I wouldn't expect this to break common scenarios, but it would be better to verify that.

> > hardcoded, because there is not minimum dimming level

> But this should not be hard coded.

> I think the minimumDimmingLevel field must be available in the initial full state resource Dimming.minimumDimmingLevel in all cases -- including the example that you have observed.

As I already stated above, there is no minimum dimming level available, so it would need to be hardcoded for this scenario.

If minimum dimming level is available, do you have any cases to investigate? As otherwise we would be solving an issue we don't know if exists, and using your own wording, that would be perverse?

Reply:
I am not sure if this is what you are looking for, but the following is the case for a normal color temperature light which had been on at a high color temperature, and then turned off via a button; then the binding was disabled and re-enabled (showing the first initial state resource); then the light was turned on via the same button to a low color temperature (showing the SSE event)..

**1. Initial Full Resource**
```
[
    {
        "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
        "id_v1": "/lights/10",
        "owner": {
            "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
            "rtype": "device"
        },
        "metadata": {
            "name": "Table Lamp A",
            "archetype": "table_shade",
            "function": "functional"
        },
        "product_data": {
            "function": "functional"
        },
        "identify": {
        },
        "on": {
            "on": false
        },
        "dimming": {
            "brightness": 61.66,
            "min_dim_level": 1.0
        },
        "dimming_delta": {
        },
        "color_temperature": {
            "mirek": 153,
            "mirek_valid": true,
            "mirek_schema": {
                "mirek_minimum": 153,
                "mirek_maximum": 454
            }
        },
        "color_temperature_delta": {
        },
        "dynamics": {
            "status": "none",
            "status_values": [
                "none"
            ],
            "speed": 0.0,
            "speed_valid": false
        },
        "alert": {
            "action_values": [
                "breathe"
            ]
        },
        "signaling": {
            "signal_values": [
                "no_signal",
                "on_off"
            ]
        },
        "mode": "normal",
        "powerup": {
            "preset": "safety",
            "configured": true,
            "on": {
                "mode": "on",
                "on": {
                    "on": true
                }
            },
            "dimming": {
                "mode": "dimming",
                "dimming": {
                    "brightness": 100.0
                }
            },
            "color": {
                "mode": "color_temperature",
                "color_temperature": {
                    "mirek": 366
                }
            }
        },
        "type": "light"
    }
]
```

**2. SSE Partial Resource**
```
[
    {
        "creationtime": "2023-12-02T14:12:47Z",
        "data": [
            {
                "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
                "id_v1": "/lights/10",
                "on": {
                    "on": true
                },
                "owner": {
                    "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
                    "rtype": "device"
                },
                "type": "light"
            },
            {
                "dimming": {
                    "brightness": 35.18
                },
                "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
                "id_v1": "/lights/10",
                "owner": {
                    "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
                    "rtype": "device"
                },
                "type": "light"
            },
            {
                "color_temperature": {
                    "mirek": 454,
                    "mirek_valid": true
                },
                "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
                "id_v1": "/lights/10",
                "owner": {
                    "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
                    "rtype": "device"
                },
                "type": "light"
            }
        ],
        "id": "53741602-d7e5-43b6-b5bc-527b1a8486b3",
        "type": "update"
    }
]
```

Reply:
PS and in any case look at the `Dimming.DEFAULT_MINIMUM_DIMMIMG_LEVEL` constant :)

Reply:
That scenario will not be influenced by the changes, because you receive **on** and **dimming** in both messages, and at no time is there any inconsistency between them.

I have one new thought though: Instead of looking at a static state and interpreting that (on + dimming 0 = on), perhaps we should actually look at the "direction". In theory both these scenarios are possible (in practice I have only seen the first):
- Current state: off, 0.
- Message: on
- Result: on + dimming > 0 because on.

And:
- Current state: on, 100.
- Message: 0
- Result: on + dimming > 0 because on.

For the second scenario it is more likely to mean "off", because previously brightness was 100, but now it's 0. So even if we didn't receive off (yet), it seems plausible it's now off.

WDYT?

I'm wondering how to best achieve this. My first thought was to remove the current fix and test coverage and instead add a method for manipulating the received resource. So that:
- If we receive on without dimming and brightness is currently 0, we explicitly add a new Dimming object with a value > 0 (as we are still discussing).
- If we receive dimming > 0 without on and on is currently off, we explicitly add a new OnState object with value on.
- If we receive off without dimming and brightness is currently > 0, we explicitly add a new Dimming object with a value of 0.
- If we receive dimming 0 without on and on is currently on, we explicitly add a new OnState object with value off.

My only concern here is that this artificial/manipulated data would then end up in the cache, which would no longer represent the actual received state. The only way to avoid this, but still react on the "direction", would be to add the logic to a method only reading the resources and then provide this information somehow to `updateChannels()` so it could be used here. Any ideas welcome. :)

Reply:
Ok. If the Thing does not have a dimmingRange property then the bug has nothing to do with the contents, or timing, of the SSE event resources. The bug is that the Initial Full resource does not contain Dimming field.

I say 'bug' because this is indeed a real bug for this particular lamp. Fundamentally the SSE events contain a 'delta' from the initial state. But if the initial state is not fully populated then the 'delta' will always be wrong.

So we should not NOT be trying to fix the data in the SSE events, but instead we should apply a patch to fix the basis of the delta. i.e. patch the initial full state resource. So I propose that in the handler, (similar to the `applyDeviceSpecificWorkArounds()`), you should add a method that does the following to fix the initial full state resource.

```
if (! hasSparseData && dimming == null) {
  dimming = new Dimming().setMinimumDimmingLevel(Dimming.DEFAULT_MINIMUM_DIMMIMG_LEVEL);
}
```

Reply:
I have no problem with that, it seems like a good idea and this would then obsolete this logic:

https://github.com/openhab/openhab-addons/blob/f0c8defabb6d7237a8ebb72686428b6ac8645ff3/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2ThingHandler.java#L402

since we would then always have a dimming level present, either actual or defaulted.

However, this is only a side-topic.

Reply:
> we would then always have a dimming level present, either actual or defaulted.

Indeed.

.. and potentially the same would apply to gamut and mirek-schema. However, this is **also** only a side-topic. :)

Reply:
Hmm, actually I think the defaulting logic of minimum dimming should be "on the fly" rather than cached. Otherwise it could cause problem with not overwriting it in cache later if it "appears late". And it would also cause it to be shown as a property, which I don't think it should when it is actually not available. It should be visible in properties that there is no actual minimum dimming declared.

Reply:
Maybe let's back to the main topic - just in case you missed this: https://github.com/openhab/openhab-addons/pull/15905#discussion_r1412816234 :)

Reply:
> defaulting logic of minimum dimming should be "on the fly" rather than cached

I don't recall the full details, but the bridge runs a scheduled task `updateSelf()` periodically to do a full state update on all things. So this 'patch' would be called in there somewhere..

Reply:
Just to be clear, that's a side-topic and will not fix the issue fixed by 52ea4f2e7bd224992c8d883c17223124251a50f9. The problem is that the SSE event contains only **on**, but not **dimming**. Dimming will either arrive in a delayed event, or not at all. So at the moment when **on** is received and current cached brightness (either from initial state or last SEE update) is 0, we have inconsistency.

Reply:
> `brightness = current cached brightness != 0 ? current cached brightness : minimum dimming level;`

But do this conversion in the thing handler (where you can access the cache) rather than where it is currently located in the resource.

Reply:
Yes, but we are back to:
- https://github.com/openhab/openhab-addons/pull/15905#discussion_r1412816234 (to not only consider current state, but also its context/"direction")
- https://github.com/openhab/openhab-addons/pull/15905#discussion_r1412792663 (minimum dimming vs. 100% - while remembering this is a specific scenario unlike your "normal" scenario, which will not be affected)

Reply:
The 'direction' and/or the choice 'minimum dimming vs. 100%' are both/either only wild guesses, and not based on any facts or received data. Whereas cached value is definitely a fact. And if no later event comes then cached value remains a fact. If cached value is 0 (yet state is on) the Hue app uses minimum dimming value. => So follow the same logic.

Reply:
> If we receive on without dimming and brightness is currently 0

Use minimum dimming

> If we receive dimming > 0 without on and on is currently off

It is off, until proven otherwise

> If we receive off without dimming and brightness is currently > 0

It is off

> If we receive dimming 0 without on and on is currently on

Does this ever happen? I doubt it..

Reply:
> > If we receive dimming > 0 without on and on is currently off
> 
> It is off, until proven otherwise

By definition in openHAB, if brightness > 0, it is **on**. So if the newly received value is **dimming** > 0, then it should be **on**, no matter what the previous **on** value was. It would be inconsistent to disregard this information.

Reply:
> By definition in openHAB, if brightness > 0, it is on

I think you are getting tired, perhaps take a break to think carefully about this.
- Your above statement does indeed apply when OH sends an HSBType command with B > 0. This should certainly send an On command to the light. (And it does).
- It does NOT apply the other way round. If the bridge sends Dimming.brightness 0 to the binding, the lamp is still on -- albeit at very low brightness i.e. minimum level -- but it is still On!
- You shall not mis-interpret Dimming.brightness 0 events as an Off state report. You shall wait until the bridge specifically sends an Off state.
- If you a receive Dimming.brightness 0 event and have not (yet) received an Off, it is ok to mess with the B part of the returned HSBType to satisfy OH (set B to min dim level). But is is not Ok to change Dimming.brightness to anything other than that which was received.
- If you receive an Off event (and have or have not received a Dimming.brightness 0 value), then it is ok to set the B part of the returned HSBType to 0 to satisfy OH. But is is not Ok to change Dimming.brightness to anything other than that which was received.

Reply:
Since there was some misunderstanding [here](https://github.com/openhab/openhab-addons/issues/15700#issuecomment-1835676939) and we do not have consensus or an implementation, I have removed disputable commit from this PR to not prolong the review process as I think the part we do agree upon wrt the merge algorithm is more or less in a ready state.

I have extracted the remaining issue to #15995.